src/distributions/uniform.rs

```diff
@@ -534,59 +533,110 @@ macro_rules! uniform_float_impl {
+                        scale = <$ty>::from_bits(scale.to_bits() - 1);
+                    }
+                } else if high != low {
+                    // No need to loop here since even if `scale` is strictly not as big
+                    // as it could be, we'll still generate good values.
+                    // But give it one try to increase our chances that
+                    // Uniform::new_inclusive is actually inclusive of the high value.
+                    let new_scale = <$ty>::from_bits(scale.to_bits() + 1);
+                    if new_scale * max_rand + low <= high {
+                        scale = new_scale;
+                    }
                 }
+
+                debug_assert!(scale >= 0.0);
+
+                UniformFloat { low, scale }
             }
 
             fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Self::X {
                 // Generate a value in the range [1, 2)
                 let value1_2 = (rng.$next_u() >> $bits_to_discard)
                                .into_float_with_exponent(0);
+
+                // Get a value in the range [0, 1) in order to avoid
+                // overflowing into infinity when multiplying with scale
+                let value0_1 = value1_2 - 1.0;
+
                 // We don't use `f64::mul_add`, because it is not available with
                 // `no_std`. Furthermore, it is slower for some targets (but
                 // faster for others). However, the order of multiplication and
                 // addition is important, because on some platforms (e.g. ARM)
                 // it will be optimized to a single (non-FMA) instruction.
-                value1_2 * self.scale + self.offset
+                value0_1 * self.scale + self.low
             }
 
             fn sample_single<R: Rng + ?Sized>(low: Self::X,
                                               high: Self::X,
                                               rng: &mut R) -> Self::X {
                 assert!(low < high,
                         "Uniform::sample_single called with low >= high");
-                let scale = high - low;
-                let offset = low - scale;
-                // Generate a value in the range [1, 2)
-                let value1_2 = (rng.$next_u() >> $bits_to_discard)
-                               .into_float_with_exponent(0);
-                // Doing multiply before addition allows some architectures to
-                // use a single instruction.
-                value1_2 * scale + offset
+                let mut scale = high - low;
+
+                loop {
+                    // Generate a value in the range [1, 2)
+                    let value1_2 = (rng.$next_u() >> $bits_to_discard)
+                                   .into_float_with_exponent(0);
+
+                    // Get a value in the range [0, 1) in order to avoid
+                    // overflowing into infinity when multiplying with scale
+                    let value0_1 = value1_2 - 1.0;
```

Comment:
Again, you're dropping a bit of precision everywhere. So why not do `value1_2 * scale + offset` as before, then at the end of the loop check whether `res` is finite and panic if not?

Reply:
I might be missing something, but I see this only costing performance, not any precision. But do let me know if I'm wrong?

In this case it should really cost very little performance though. While we need an additional subtraction here, we are also saving a subtraction since we're not calculating `offset`. Though for constant inputs, `offset` and `scale` can be computed at compile time. But arguably someone caring about performance and using constant ranges should create a `Uniform`.

Additionally, using the `offset` approach means that we have to check the resulting value both against `high` and against `low`, so the resulting code is slower. Here's the perf numbers I get:

Using `value1_2 * scale + offset`
```
test gen_range_f32               ... bench:       4,835 ns/iter (+/- 857) = 827 MB/s
test gen_range_f64               ... bench:       6,896 ns/iter (+/- 641) = 1160 MB/s
```

Using `value0_1 * scale + low`
```
test gen_range_f32               ... bench:       4,335 ns/iter (+/- 251) = 922 MB/s
test gen_range_f64               ... bench:       5,712 ns/iter (+/- 916) = 1400 MB/s
```